etcdmain/config.go

```diff
@@ -317,6 +356,141 @@ func (cfg *config) Parse(arguments []string) error {
+	b, err := ioutil.ReadFile(cfg.configFile)
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal the YAML.
+	err = yaml.Unmarshal(b, cfg)
+	if err != nil {
+		return err
+	}
+
+	// Make sure the defaults are set for the missing fields.
+	cfg.setDefaultsIfNeeded()
+
+	// Validate whether conflicting parameters are specified.
+	nSet := 0
+	for _, s := range []string{cfg.Durl, cfg.InitialCluster, cfg.DnsCluster} {
+		if s != "" {
+			nSet += 1
+		}
+	}
+	if nSet > 1 {
+		return ErrConflictBootstrapFlags
+	}
+
+	// Read in listen-peer-urls.
+	u, err := types.NewURLs(strings.Split(cfg.LPUrlsCfgFile, ","))
+	if err != nil {
+		plog.Fatalf("unexpected error setting up listen-peer-urls: %v", err)
+	}
+	cfg.lpurls = []url.URL(u)
+
+	// Read in listen-client-urls.
+	u, err = types.NewURLs(strings.Split(cfg.LCUrlsCfgFile, ","))
+	if err != nil {
+		plog.Fatalf("unexpected error setting up listen-client-urls: %v", err)
+	}
+	cfg.lcurls = []url.URL(u)
+
+	// Read in cors.
+	if err := cfg.corsInfo.Set(cfg.CorsCfgFile); err != nil {
+		plog.Panicf("unexpected error setting up cors: %v", err)
+	}
+
+	// Read in intial-advertise-peer-urls.
+	u, err = types.NewURLs(strings.Split(cfg.ApurlsCfgFile, ","))
+	if err != nil {
+		plog.Fatalf("unexpected error setting up initial-advertise-peer-urls: %v", err)
+	}
+	cfg.apurls = []url.URL(u)
+
+	// Read in advertise-client-urls.
+	u, err = types.NewURLs(strings.Split(cfg.AcurlsCfgFile, ","))
+	if err != nil {
+		plog.Fatalf("unexpected error setting up advertise-peer-urls: %v", err)
+	}
+	cfg.acurls = []url.URL(u)
+
+	// Read in cluster-state.
+	if err := cfg.clusterState.Set(cfg.ClusterStateCfgFile); err != nil {
+		plog.Panicf("unexpected error setting up clusterStateFlag: %v", err)
+	}
+
+	// Read in discovery-fallback.
+	if err := cfg.fallback.Set(cfg.FallbackCfgFile); err != nil {
+		plog.Panicf("unexpected error setting up discovery-fallback flag: %v", err)
+	}
+
+	// Read in proxy.
+	if err := cfg.proxy.Set(cfg.ProxyCfgFile); err != nil {
+		plog.Panicf("unexpected error setting up proxyFlag: %v", err)
+	}
+
+	// Copy over the security details.
+	copySecurityDetails := func(tls *transport.TLSInfo, ysc *securityConfig) {
+		tls.CAFile = ysc.CAFile
+		tls.CertFile = ysc.CertFile
+		tls.KeyFile = ysc.KeyFile
+		tls.ClientCertAuth = ysc.CertAuth
+		tls.TrustedCAFile = ysc.TrustedCAFile
+	}
+	copySecurityDetails(&cfg.clientTLSInfo, &cfg.ClientSecurityCfgFile)
+	copySecurityDetails(&cfg.peerTLSInfo, &cfg.PeerSecurityCfgFile)
+
+	// When etcd runs in member mode user needs to set --advertise-client-urls if --listen-client-urls is set.
+	mayFallbackToProxy := cfg.Durl != "" && cfg.fallback.String() == fallbackFlagProxy
+	mayBeProxy := cfg.proxy.String() != proxyFlagOff || mayFallbackToProxy
+	if !mayBeProxy {
+		if cfg.LCUrlsCfgFile != "" && cfg.AcurlsCfgFile == "" {
+			return errUnsetAdvertiseClientURLsFlag
+		}
+	}
+
+	return nil
+}
+
+func (cfg *config) Parse(arguments []string) error {
```

Comment:
move this func before the two helper funcs (cfg.configFromFile and cfg.configFromCmdLine)?